### PR TITLE
M3-782 IP Sharing 

### PR DESCRIPTION
**Pull request by asauber**
* Implement IP Sharing Panel
* Show shared IPs in Linode's IPv4 table
* Allow updating and deleting shared IPs

### CODE REVIEW

**Comment by martinmckenna** [comment]
```
Warning: Can't call setState (or forceUpdate) on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in IPSharingPanel (created by WithStyles(IPSharingPanel))
```

will need to do `is mounted` checking

**Comment by martinmckenna** [comment]
screen shot 2018-06-29 at 4 02 32 pm

I wouldn't say this is blocking merge, but it would be nice if the IPs were removed from the list when they are selected

screen shot 2018-06-29 at 3 54 56 pm

I also think there should be an empty state explaining what the user needs to do in order to see values in the dropdown

File: src/features/linodes/LinodesDetail/LinodeNetworking/IPSharingPanel.tsx [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+            fullWidth={false}
+            className={classes.ipField}
+          >
+            {ipChoices.map((ipChoice: string, choiceIdx: number) =>
+              <MenuItem 
+                data-ip-idx={idx}
+                key={choiceIdx}
+                value={ipChoice}
+              >
+                {ipChoice}
```
Comment (martinmckenna) [inline src/features/linodes/LinodesDetail/LinodeNetworking/IPSharingPanel.tsx]: possibly add linode label here as well?

Comment (Jskobos) [reply src/features/linodes/LinodesDetail/LinodeNetworking/IPSharingPanel.tsx]: I agree, label (in parentheses maybe) would be helpful.

File: src/features/linodes/LinodesDetail/LinodeNetworking/IPSharingPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+      .catch((response) => {
+        const errors = pathOr([], ['response', 'data', 'errors'], response);
+        this.setState({
+          errors,
+          loading: false,
+        })
+      })
+  }
+
+  componentWillReceiveProps(nextProps: Props) {
```
Comment (Jskobos) [inline src/features/linodes/LinodesDetail/LinodeNetworking/IPSharingPanel.tsx]: It's way too early to start being picky about this, but this method is deprecated in favor of getDerivedStateFromProps. 

Also, it's probably better make the state update conditional (or remove it, not sure it's necessary). Not a big deal in this case, but if you select a new IP address to share then open one of the drawers on the page before pressing save, the selection disappears.

**Comment by rjstires** [comment]
@asauber do you need additional review on this, or should I wait for your response to Marty's feedback?

**Comment by asauber** [comment]
@rjstires I'll ping you for review after addressing the current feedback.

**Comment by asauber** [comment]
@martinmckenna @Jskobos @rjstires Ready for re-review

File: src/services/networking.ts [hunk 3]
```diff
@@ -16,10 +16,21 @@ export const assignAddresses = (data: any) =>
     setData(data),
   );
 
-export const getIPs = () => {
-  return Request<Linode.ResourcePage<Linode.IPAddress>>(
+export const getIPs = (region?: string) => {
```
Comment (rjstires) [inline src/services/networking.ts]: Not a request for change, but an alternative would be;
```
export const getIPs = (region?: string) =>
  Request(
    setURL(`${API_ROOT}/networking/ips`),
    setMethod('GET'),
    when(Boolean(region), setXFilter({ region })),
  )
    .then(response => response.data);
```

File: src/features/linodes/LinodesDetail/LinodeNetworking/IPSharingPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+        </Grid>
+        <Grid item>
+          <TextField disabled value={ip} className={classes.ipField} />
+        </Grid>
+      </Grid>
+    );
+  }
+
+  onIPSelect = (e: React.ChangeEvent<HTMLSelectElement>) => {
+    const ipIdx = e.currentTarget.getAttribute('data-ip-idx');
```
Comment (rjstires) [inline src/features/linodes/LinodesDetail/LinodeNetworking/IPSharingPanel.tsx]: I tested this pretty thoroughly in the `mui-performance-test` project. Using a curried function does not result in additional re-renders, and using getAttribute is worse render performance.

**Pull request merged**